File: Java-Core/src/main/java/presto/declaration/CategoryDeclaration.java
package presto.declaration;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.grammar.MethodDeclarationList;
import presto.grammar.Identifier;
import presto.parser.IProblemListener;
import presto.runtime.Context;
import presto.type.CategoryType;
import presto.type.IType;
import presto.utils.CodeWriter;
import presto.utils.IdentifierList;
import presto.value.IInstance;

public abstract class CategoryDeclaration extends BaseDeclaration {
	
	IdentifierList attributes;
	
	public CategoryDeclaration(Identifier name) {
		super(name);
	}

	public CategoryDeclaration(Identifier name, IdentifierList attributes) {
		super(name);
		this.attributes = attributes;
	}

	public void setAttributes(IdentifierList attributes) {
		this.attributes = attributes;
	}
	
	public IdentifierList getAttributes() {
		return attributes;
	}
		
	@Override
	public void register(Context context) throws SyntaxError {
		context.registerDeclaration(this);
	}
	
	@Override
	public IType check(Context context) throws SyntaxError {
		if(attributes!=null) for(Identifier attribute : attributes) {
			AttributeDeclaration ad = context.getRegisteredDeclaration(AttributeDeclaration.class, attribute);
			if(ad==null) {
				IProblemListener pl = context.getProblemListener();
				if(pl!=null)
					pl.reportUnknownAttribute(attribute.toString(), attribute);
				else
					throw new SyntaxError("Unknown attribute: \"" + attribute + "\"");
			}
		}
		return new CategoryType(this.getIdentifier());
	}
	
	@Override
	public CategoryType getType(Context context) {
		return new CategoryType(getIdentifier());
	}

	public boolean hasAttribute(Context context, Identifier name) {
		 return attributes!=null && attributes.contains(name);
	}

	public boolean hasMethod(Context context, String key, Object object) {
		return false;
	}

	public boolean isDerivedFrom(Context context, CategoryType categoryType) {
		return false;
	}

	public IdentifierList getDerivedFrom() {
		return null;
	}

	public abstract IInstance newInstance() throws PrestoError;

	public void checkConstructorContext(Context context) throws SyntaxError {
		// nothing to do
	}
	
	@Override
	public void toDialect(CodeWriter writer) {
		writer = writer.newInstanceWriter(getType(writer.getContext()));
		switch(writer.getDialect()) {
		case E:
			toEDialect(writer);
			break;
		case O:
			toODialect(writer);
			break;
		case S:
			toSDialect(writer);
			break;
		}
	}

	protected abstract void toEDialect(CodeWriter writer);

	protected void protoToEDialect(CodeWriter writer, boolean hasMethods, boolean hasMappings) {
		boolean hasAttributes = attributes!=null && attributes.size()>0;
		writer.append("define ");
		writer.append(getName());
		writer.append(" as: ");
		categoryTypeToEDialect(writer);
		if(hasAttributes) {
			if(attributes.size()==1)
				writer.append(" with attribute: ");
			else
				writer.append(" with attributes: ");
			attributes.toDialect(writer, true);
		}
		if(hasMethods) {
			if(hasAttributes)
				writer.append(", and methods:");
			else 
				writer.append(" with methods:");
		} else if (hasMappings) {
			if(hasAttributes)
				writer.append(", and bindings:");
			else 
				writer.append(" with bindings:");
		}
		writer.newLine();	
	}
	
	protected void methodsToEDialect(CodeWriter writer, MethodDeclarationList methods) {
		writer.indent();
		for(IDeclaration decl : methods) {
			writer.newLine();
			CodeWriter w = writer.newMemberWriter();
			decl.toDialect(w);
		}
		writer.dedent();
	}

	protected void methodsToODialect(CodeWriter writer, MethodDeclarationList methods) {
		for(IDeclaration decl : methods) {
			CodeWriter w = writer.newMemberWriter();
			decl.toDialect(w);
			w.newLine();
		}
	}



	protected abstract void categoryTypeToEDialect(CodeWriter writer);

	protected abstract void toODialect(CodeWriter writer);
	
	protected void toODialect(CodeWriter writer, boolean hasBody) {
		categoryTypeToODialect(writer);
		writer.append(" ");
		writer.append(getName());
		if(attributes!=null) {
			writer.append('(');
			attributes.toDialect(writer, true);
			writer.append(')');
		}	
		categoryExtensionToODialect(writer);
		if(hasBody) {
			writer.append(" {\n");
			writer.newLine();
			writer.indent();
			bodyToODialect(writer);
			writer.dedent();
			writer.append('}');
			writer.newLine();
		} else
			writer.append(';');
	}

	protected abstract void categoryTypeToODialect(CodeWriter writer);

	protected void categoryExtensionToODialect(CodeWriter writer) {
		// by default no extension
	}

	protected abstract void bodyToODialect(CodeWriter writer);

	protected abstract void toSDialect(CodeWriter writer);

	protected void protoToPDialect(CodeWriter writer, IdentifierList derivedFrom) {
		categoryTypeToPDialect(writer);
		writer.append(" ");
		writer.append(getName());
		writer.append("(");
		if(derivedFrom!=null) {
			derivedFrom.toDialect(writer, false);
			if(attributes!=null)
				writer.append(", ");
		}
		if(attributes!=null)
			attributes.toDialect(writer, false);
		writer.append("):\n");
		writer.newLine();
	}

	protected abstract void categoryTypeToPDialect(CodeWriter writer);
	
	




}

File: Java-CodeServer/src/main/resources/js/prompto/declaration/CategoryDeclaration.js
var BaseDeclaration = require("./BaseDeclaration").BaseDeclaration;
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var CategoryType = require("../type/CategoryType").CategoryType;

function CategoryDeclaration(name, attributes) {
	BaseDeclaration.call(this, name);
	this.attributes = attributes || null;
	this.derivedFrom = null;
}

CategoryDeclaration.prototype = Object.create(BaseDeclaration.prototype);
CategoryDeclaration.prototype.constructor = CategoryDeclaration;


CategoryDeclaration.prototype.register = function(context) {
	context.registerDeclaration(this);
};

CategoryDeclaration.prototype.check = function(context) {
	if(this.attributes!=null) {
		for(var i=0;i<this.attributes.length;i++) {
			var ad = context.getRegisteredDeclaration(this.attributes[i]);
			if (ad == null) {
				throw new SyntaxError("Unknown attribute: \"" + this.attributes[i] + "\"");
			}
		}
	}
	return new CategoryType(this.name);
}

CategoryDeclaration.prototype.getType = function(context) {
	return new CategoryType(this.name);
};

CategoryDeclaration.prototype.hasAttribute = function(context, name) {
	 return this.attributes!=null && this.attributes.indexOf(name)>=0;
};

CategoryDeclaration.prototype.hasMethod = function(context, key, object) {
	return false;
};

CategoryDeclaration.prototype.isDerivedFrom = function(context, categoryType) {
	return false;
};


CategoryDeclaration.prototype.checkConstructorContext = function(context) {
	// nothing to do
};

CategoryDeclaration.prototype.toDialect = function(writer) {
    var type = this.getType(writer.context);
    writer = writer.newInstanceWriter(type);
    writer.toDialect(this);
};


CategoryDeclaration.prototype.protoToEDialect = function(writer, hasMethods, hasBindings) {
    var hasAttributes = this.attributes!=null && this.attributes.length>0;
    writer.append("define ");
    writer.append(this.name);
    writer.append(" as: ");
    this.categoryTypeToEDialect(writer);
    if(hasAttributes) {
        if(this.attributes.length==1)
            writer.append(" with attribute: ");
        else
            writer.append(" with attributes: ");
        this.attributes.toDialect(writer, true);
    }
    if(hasMethods) {
        if(hasAttributes)
            writer.append(", and methods:");
        else
            writer.append(" with methods:");
    } else if (hasBindings) {
        if(hasAttributes)
            writer.append(", and bindings:");
        else
            writer.append(" with bindings:");
    }
    writer.newLine();
};


CategoryDeclaration.prototype.methodsToEDialect = function(writer, methods) {
    writer.indent();
    for(var i=0; i<methods.length; i++) {
        writer.newLine();
        var w = writer.newMemberWriter();
        methods[i].toDialect(w);
    }
    writer.dedent();
};

CategoryDeclaration.prototype.methodsToODialect = function(writer, methods) {
    for(var i =0; i<methods.length; i++) {
        var w = writer.newMemberWriter();
        methods[i].toDialect(w);
        w.newLine();
    }
}


CategoryDeclaration.prototype.allToODialect = function(writer, hasBody) {
    this.categoryTypeToODialect(writer);
    writer.append(" ");
    writer.append(this.name);
    if(this.attributes!=null) {
        writer.append('(');
        this.attributes.toDialect(writer, true);
        writer.append(')');
    }
    this.categoryExtensionToODialect(writer);
    if(hasBody) {
        writer.append(" {\n");
        writer.newLine();
        writer.indent();
        this.bodyToODialect(writer);
        writer.dedent();
        writer.append('}');
        writer.newLine();
    } else
        writer.append(';');
};

CategoryDeclaration.prototype.categoryExtensionToODialect = function(writer) {
    // by default no extension
};


CategoryDeclaration.prototype.protoToSDialect = function(writer, derivedFrom) {
    this.categoryTypeToSDialect(writer);
    writer.append(" ");
    writer.append(this.name);
    writer.append("(");
    if(this.derivedFrom!=null) {
        this.derivedFrom.toDialect(writer, false);
        if(this.attributes!=null)
            writer.append(", ");
    }
    if(this.attributes!=null)
        this.attributes.toDialect(writer, false);
    writer.append("):\n");
    writer.newLine();
};


exports.CategoryDeclaration = CategoryDeclaration;

